File: ckgame-java/src/main/java/com/ckgame/military/UnitStack.java
package com.ckgame.military;

/**
 * 兵力堆叠：同一单位类型的人数与最大人数。
 */
public final class UnitStack {
    public UnitType unitType;
    public int men;
    public int maxMen;

    public UnitStack(UnitType unitType, int men, int maxMen) {
        this.unitType = unitType;
        this.men = men;
        this.maxMen = maxMen;
    }

    /** 当前兵力占最大兵力的比例。 */
    public double strengthRatio() {
        return maxMen == 0 ? 0.0 : (double) men / maxMen;
    }

    /** 考虑兵力比例后的战斗力。 */
    public double combatPower() {
        return men * unitType.damage() * strengthRatio();
    }

    /** 承受伤亡，返回实际死亡人数。 */
    public int takeCasualties(int amount) {
        int killed = Math.min(men, amount);
        men -= killed;
        return killed;
    }
}

File: ckgame-ts/src/military/UnitStack.ts
import { UnitType, unitDamageValue } from './UnitType.js';

/**
 * A stack of units of the same type.
 */
export class UnitStack {
  public unitType: UnitType;
  public men: number;
  public maxMen: number;

  constructor(unitType: UnitType, men: number, maxMen: number) {
    this.unitType = unitType;
    this.men = men;
    this.maxMen = maxMen;
  }

  /** Current men as ratio of max. */
  strengthRatio(): number {
    return this.maxMen === 0 ? 0 : this.men / this.maxMen;
  }

  /** Combat power accounting for current strength ratio. */
  combatPower(): number {
    return this.men * unitDamageValue(this.unitType) * this.strengthRatio();
  }

  /** Take casualties, returns actual killed. */
  takeCasualties(amount: number): number {
    const killed = Math.min(this.men, amount);
    this.men -= killed;
    return killed;
  }
}
